diff --git a/features.module b/features.module
index 35dc0c6..a38ab58 100644
--- a/features.module
+++ b/features.module
@@ -977,9 +977,16 @@ function _features_restore($op, $items = array()) {
     }
   }
 
+  $components = features_get_components();
+  //Sort all components by weight
+  uasort($components, 'drupal_sort_weight');
+  $components_sorted = array_keys($components);
+
   // Invoke global pre restore hook.
   module_invoke_all('features_pre_restore', $op, $items);
   foreach ($items as $module_name => $components) {
+    //Make sure components are sorted by weight
+    $components = array_intersect($components_sorted, $components);
     // If feature is totally locked, do not execute past this stage.
     if ($lockable && features_feature_is_locked($module_name)) {
       watchdog('features', 'Tried @actioning a locked @module_name, aborted.', array('@action' => $log_action, '@module_name' => $module_name));
