diff --git a/entity_translation.admin.inc b/entity_translation.admin.inc
index c0992ee..2e87b79 100644
--- a/entity_translation.admin.inc
+++ b/entity_translation.admin.inc
@@ -411,6 +411,7 @@ function theme_entity_translation_overview_outdated($variables){
 function entity_translation_delete_confirm($form, $form_state, $entity_type, $entity, $langcode) {
   $handler = entity_translation_get_handler($entity_type, $entity);
   $handler->setFormLanguage($langcode);
+  $handler->initPathScheme();
   $languages = language_list();
 
   $form = array(
