diff --git a/entity_translation.module b/entity_translation.module
index f7ca0ab..e11d057 100644
--- a/entity_translation.module
+++ b/entity_translation.module
@@ -1274,10 +1274,14 @@ function entity_translation_field_attach_form($entity_type, $entity, &$form, &$f
 function entity_translation_prepare_element($element, &$form_state) {
   static $drupal_static_fast;
   if (!isset($drupal_static_fast)) {
-    $drupal_static_fast['source_forms'] = &drupal_static(__FUNCTION__, array());
+    $drupal_static_fast = &drupal_static(__FUNCTION__, array(
+      'source_forms' => array(),
+      'source_form_states' => array(),
+    ));
   }
 
   $source_forms = &$drupal_static_fast['source_forms'];
+  $source_form_states = &$drupal_static_fast['source_form_states'];
   $form = $form_state['complete form'];
   $build_id = $form['#build_id'];
   $source = $element['#source'];
@@ -1294,9 +1298,11 @@ function entity_translation_prepare_element($element, &$form_state) {
     $source_form_state = $form_state;
     field_attach_form($entity_type, $element['#entity'], $source_form, $source_form_state, $source);
     $source_forms[$build_id][$source][$entity_type][$id] = &$source_form;
+    $source_form_states[$build_id][$source][$entity_type][$id] = &$source_form_state;
   }
 
   $source_form = &$source_forms[$build_id][$source][$entity_type][$id];
+  $source_form_state = $source_form_states[$build_id][$source][$entity_type][$id];
   $langcode = $element['#language'];
   $field_name = $element['#field_name'];
 
@@ -1307,6 +1313,7 @@ function entity_translation_prepare_element($element, &$form_state) {
   if (isset($source_form[$field_name][$source])) {
     $element[$langcode] = $source_form[$field_name][$source];
     entity_translation_form_element_language_replace($element, $source, $langcode);
+    entity_translation_form_element_state_replace($element, $source_form[$field_name], $field_name, $source_form_state, $form_state);
     unset($element[$element['#previous']]);
   }
 
@@ -1314,6 +1321,35 @@ function entity_translation_prepare_element($element, &$form_state) {
 }
 
 /**
+ * Helper function. Sets the right values in $form_state['field'] when using
+ * source language values as defaults.
+ */
+function entity_translation_form_element_state_replace($element, $source_element, $field_name, $source_form_state, &$form_state) {
+  if (isset($source_element['#language'])) {
+    $source = $source_element['#language'];
+
+    // Iterate through the form structure recursively.
+    foreach (element_children($element) as $key) {
+      if (isset($source_element[$key])) {
+        entity_translation_form_element_state_replace($element[$key], $source_element[$key], $key, $source_form_state, $form_state);
+      }
+      elseif (isset($source_element[$source])) {
+        entity_translation_form_element_state_replace($element[$key], $source_element[$source], $key, $source_form_state, $form_state);
+      }
+    }
+
+    if (isset($source_element[$source]['#field_parents'])) {
+      $source_parents = $source_element[$source]['#field_parents'];
+      $langcode = $element['#language'];
+      $parents = $element[$langcode]['#field_parents'];
+      $source_state = field_form_get_state($source_parents, $field_name, $source, $source_form_state);
+      drupal_alter('entity_translation_source_field_state', $source_state);
+      field_form_set_state($parents, $field_name, $langcode, $form_state, $source_state);
+    }
+  }
+}
+
+/**
  * Helper function. Recursively replaces the source language with the given one.
  */
 function entity_translation_form_element_language_replace(&$element, $source, $langcode) {
