diff --git a/plugins/content_types/search_api_page_result.inc b/plugins/content_types/search_api_page_result.inc
index 3f2815f..3734887 100644
--- a/plugins/content_types/search_api_page_result.inc
+++ b/plugins/content_types/search_api_page_result.inc
@@ -107,7 +107,7 @@ function search_api_page_result_content_type_edit_form($form, &$form_state) {
     'search_api_page_result' => t('Themed as search results'),
   );
   // For entities, we also add all entity view modes.
-  if ($entity_info = entity_get_info($index->item_type)) {
+  if ($entity_info = entity_get_info($index->getEntityType())) {
     foreach ($entity_info['view modes'] as $mode => $mode_info) {
       $view_modes[$mode] = $mode_info['label'];
     }
