diff --git a/includes/SearchApiEtDatasourceController.php b/includes/SearchApiEtDatasourceController.php
index d1c9a93..9e7075f 100644
--- a/includes/SearchApiEtDatasourceController.php
+++ b/includes/SearchApiEtDatasourceController.php
@@ -446,30 +446,17 @@ class SearchApiEtDatasourceController extends SearchApiEntityDataSourceControlle
    * @return array
    */
   protected function filterTrackableIds(SearchApiIndex $index, $item_ids) {
-    if (empty($item_ids))
+    if (empty($item_ids)) {
       return array();
-
+    }
     $merged_ids = $this->getGroupedItemsIdsByEntity($item_ids);
     if (empty($merged_ids)) {
       return array();
     }
+    $trackable_item_ids = $this->getTrackableItemIds($index, array_keys($merged_ids));
 
-    $ret = array();
-    $ids = $this->getTrackableEntityIds($index, array_keys($merged_ids));
-
-    if (!empty($ids)) {
-      // Keeping only the Entity IDs matched by the query
-      $item_ids = array_intersect_key($merged_ids, array_flip($ids));
-
-      // Rebuilding the $ret to contain the '{entity_id}/{language}' IDs
-      if (!empty($item_ids)) {
-        foreach ($item_ids as $ids) {
-          $ret = array_merge($ret, $ids);
-        }
-      }
-    }
-
-    return $ret;
+    // Keeping only the Entity IDs matched by the query
+    return array_intersect($item_ids, $trackable_item_ids);
   }
 
   /**
